File: examples/standalone/ts/05-unions.ts
// Example 5: Functions with enum/union parameters
// Tests shatter's ability to enumerate discriminated union variants.

type Shape =
    | { kind: "circle"; radius: number }
    | { kind: "rectangle"; width: number; height: number }
    | { kind: "triangle"; base: number; height: number };

// EXPECTED BRANCHES:
//   1. kind === "circle"    AND radius <= 0    → throws Error("non-positive radius")
//   2. kind === "circle"    AND radius > 0     → returns π * r²
//   3. kind === "rectangle" AND w/h <= 0       → throws Error("non-positive dimension")
//   4. kind === "rectangle" AND w/h > 0        → returns w * h
//   5. kind === "triangle"  AND base/h <= 0    → throws Error("non-positive dimension")
//   6. kind === "triangle"  AND base/h > 0     → returns 0.5 * base * h

export function computeArea(shape: Shape): number {
    switch (shape.kind) {
        case "circle":
            if (shape.radius <= 0) {
                throw new Error("non-positive radius");
            }
            return Math.PI * shape.radius * shape.radius;

        case "rectangle":
            if (shape.width <= 0 || shape.height <= 0) {
                throw new Error("non-positive dimension");
            }
            return shape.width * shape.height;

        case "triangle":
            if (shape.base <= 0 || shape.height <= 0) {
                throw new Error("non-positive dimension");
            }
            return 0.5 * shape.base * shape.height;
    }
}

type HttpMethod = "GET" | "POST" | "PUT" | "DELETE";

interface ApiRequest {
    method: HttpMethod;
    path: string;
    body?: string;
    authenticated: boolean;
}

// EXPECTED BRANCHES:
//   1. path is empty                          → throws Error("empty path")
//   2. method === "GET"                       → "read"
//   3. method === "DELETE" AND !authenticated  → throws Error("auth required")
//   4. method === "DELETE" AND authenticated   → "delete"
//   5. method === "POST"  AND no body         → throws Error("body required")
//   6. method === "POST"  AND has body        → "create"
//   7. method === "PUT"   AND no body         → throws Error("body required")
//   8. method === "PUT"   AND has body        → "update"

export function routeRequest(req: ApiRequest): string {
    if (req.path.length === 0) {
        throw new Error("empty path");
    }

    switch (req.method) {
        case "GET":
            return "read";

        case "DELETE":
            if (!req.authenticated) {
                throw new Error("auth required");
            }
            return "delete";

        case "POST":
            if (!req.body) {
                throw new Error("body required");
            }
            return "create";

        case "PUT":
            if (!req.body) {
                throw new Error("body required");
            }
            return "update";
    }
}

File: examples/standalone/go/05-unions.go
package main

import (
	"fmt"
	"math"
	"strings"
)

// Example 5: Interface-based dispatch (Go equivalent of discriminated unions).
// Tests shatter's ability to enumerate type variants via interface + type switch.

// Shape is implemented by Circle, RectangleShape, and Triangle.
type Shape interface {
	Kind() string
}

type Circle struct{ Radius float64 }
type RectangleShape struct{ Width, Height float64 }
type Triangle struct{ Base, Height float64 }

func (c Circle) Kind() string         { return "circle" }
func (r RectangleShape) Kind() string  { return "rectangle" }
func (t Triangle) Kind() string        { return "triangle" }

// ComputeArea — 6 branches: circle+radius≤0→error, circle→πr²,
// rectangle+dim≤0→error, rectangle→w*h, triangle+dim≤0→error, triangle→0.5*b*h.
// Analyzer should detect type switch arms and nested dimension validation.
func ComputeArea(shape Shape) (float64, error) {
	switch s := shape.(type) {
	case Circle:
		if s.Radius <= 0 {
			return 0, fmt.Errorf("non-positive radius")
		}
		return math.Pi * s.Radius * s.Radius, nil
	case RectangleShape:
		if s.Width <= 0 || s.Height <= 0 {
			return 0, fmt.Errorf("non-positive dimension")
		}
		return s.Width * s.Height, nil
	case Triangle:
		if s.Base <= 0 || s.Height <= 0 {
			return 0, fmt.Errorf("non-positive dimension")
		}
		return 0.5 * s.Base * s.Height, nil
	default:
		return 0, fmt.Errorf("unknown shape")
	}
}

// ApiRequest represents an HTTP API request with optional body.
type ApiRequest struct {
	Method        string
	Path          string
	Body          string
	HasBody       bool
	Authenticated bool
}

// RouteRequest — 8 branches: empty path→error, GET→"read",
// DELETE+!auth→error, DELETE+auth→"delete", POST+no body→error, POST→"create",
// PUT+no body→error, PUT→"update".
// Analyzer should detect string-equality dispatch on Method with nested checks.
func RouteRequest(req ApiRequest) (string, error) {
	if len(req.Path) == 0 {
		return "", fmt.Errorf("empty path")
	}

	switch strings.ToUpper(req.Method) {
	case "GET":
		return "read", nil
	case "DELETE":
		if !req.Authenticated {
			return "", fmt.Errorf("auth required")
		}
		return "delete", nil
	case "POST":
		if !req.HasBody {
			return "", fmt.Errorf("body required")
		}
		return "create", nil
	case "PUT":
		if !req.HasBody {
			return "", fmt.Errorf("body required")
		}
		return "update", nil
	default:
		return "", fmt.Errorf("unknown method")
	}
}
